libbeat/metric/system/cgroup/util.go

```diff
@@ -0,0 +1,293 @@
+		line := strings.TrimSpace(sc.Text())
+		if line == "" {
+			continue
+		}
+
+		mount, err := parseMountinfoLine(line)
+		if err != nil {
+			return Mountpoints{}, err
+		}
+
+		if !strings.HasPrefix(mount.mountpoint, rootfsMountpoint) {
+			continue
+		}
+
+		// cgroupv1 option
+		if mount.filesystemType == "cgroup" {
+			for _, opt := range mount.superOptions {
+				// Sometimes the subsystem name is written like "name=blkio".
+				fields := strings.SplitN(opt, "=", 2)
+				if len(fields) > 1 {
+					opt = fields[1]
+				}
+
+				// Test if option is a subsystem name.
+				if _, found := subsystems[opt]; found {
+					// Add the subsystem mount if it does not already exist.
+					if _, exists := mounts[opt]; !exists {
+						mounts[opt] = mount.mountpoint
+					}
+				}
+			}
+		}
+
+		// V2 option
+		if mount.filesystemType == "cgroup2" {
+			mountInfo.V2Loc = mount.mountpoint
+		}
+
+	}
+
+	mountInfo.V1Mounts = mounts
+
+	return mountInfo, sc.Err()
+}
+
+// ProcessCgroupPaths returns the cgroups to which a process belongs and the
+// pathname of the cgroup relative to the mountpoint of the subsystem.
+func (r Reader) ProcessCgroupPaths(pid int) (map[string]ControllerPath, error) {
+	cgroup, err := os.Open(filepath.Join(r.rootfsMountpoint, "proc", strconv.Itoa(pid), "cgroup"))
+	if err != nil {
+		return nil, err //return a blank error so other events can use any file not found errors
+	}
+	defer cgroup.Close()
+
+	cPaths := map[string]ControllerPath{}
+	sc := bufio.NewScanner(cgroup)
+	for sc.Scan() {
+		// http://man7.org/linux/man-pages/man7/cgroups.7.html
+		// Format: hierarchy-ID:subsystem-list:cgroup-path
+		// Example:
+		// 2:cpu:/docker/b29faf21b7eff959f64b4192c34d5d67a707fe8561e9eaa608cb27693fba4242
+		line := sc.Text()
+
+		fields := strings.Split(line, ":")
+		if len(fields) != 3 {
+			continue
+		}
+
+		path := fields[2]
+		if r.cgroupsHierarchyOverride != "" {
+			path = r.cgroupsHierarchyOverride
+		}
+		// cgroup V2
+		// cgroup v2 controllers will always start with this string
+		if strings.Contains(line, "0::/") {
+			// if you're running inside a container
+			// that's operating with a hybrid cgroups config,
+			// the containerized process won't see the V2 mount
+			// inside /proc/self/mountinfo if docker is using cgroups V1
+			// For this very annoying edge case, revert to the hostfs flag
+			// If it's not set, warn the user that they've hit this.
+			controllerPath := filepath.Join(r.cgroupMountpoints.V2Loc, path)
+			// Depending on the test environment, Hostfs can either be blank, or `/`
+			if r.cgroupMountpoints.V2Loc == "" && len(paths.Paths.Hostfs) <= 1 {
+				logp.L().Infof(`PID %d contains a cgroups V2 path (%s) but no V2 mountpoint was found.
+This may be because metricbeat is running inside a container on a hybrid system.
+To monitor cgroups V2 processess in this way, mount the unified (V2) hierarchy inside
+the container as /sys/fs/cgroup/unified and start metricbeat with --system.hostfs.`, pid, line)
```

Comment:
@fearful-symmetry in an elastic-agent started with `elastic-package stack` in my laptop, I see continuously (every collection period) this message, I think we should move it to the debug level. In a system when this happens this is going to be continuosly logged.

I use a quite vanilla Ubuntu 18.04.5 LTS, with Linux 4.15, and docker 20.10.7. So maybe this is not so an edge case.

Also, as far as I know, both my host and the docker containers are using cgroups V1 (I don't see any cgroups2 mountpoint or any unified cgroup) so I wonder if this comment about hybrid systems can be misleading in this case.

For reference, this is what I see in logs:
```
2021-08-18T08:15:09.797Z	INFO	cgroup/util.go:257	PID 10 contains a cgroups V2 path (0::/system.slice/containerd.service) but no V2 mountpoint was found.
This may be because metricbeat is running inside a container on a hybrid system.
To monitor cgroups V2 processess in this way, mount the unified (V2) hierarchy inside
the container as /sys/fs/cgroup/unified and start metricbeat with --system.hostfs.
```

This is how `/proc/10/cgroup` looks like in the container:
```
12:freezer:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
11:devices:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
10:cpu,cpuacct:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
9:pids:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
8:hugetlb:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
7:blkio:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
6:perf_event:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
5:cpuset:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
4:memory:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
3:net_cls,net_prio:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
2:rdma:/
1:name=systemd:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
0::/system.slice/containerd.service
```

Reply:
@jsoriano 
> `0::/system.slice/containerd.service`

So, that's the V2 controller there. You said you didn't see any unified mountpoint on the system? Can you get me `/proc/self/mountinfo` ?

However, I think you're right that we might want some kind of behavior to prevent this message from getting spammy. Unfortunately, from within a docker container it's really hard to get that info without the user mounting `sysfs` inside the container. Maybe _just_ have it print the message at startup?

Reply:
Oh, you are right, there is some `cgroup2` around, I didn't see it.
```
$ cat /proc/self/mountinfo  | grep cgroup2
34 33 0:29 / /sys/fs/cgroup/unified rw,nosuid,nodev,noexec,relatime shared:10 - cgroup2 cgroup rw,nsdelegate
```

So yeah, let's keep the message but trying to make it less verbose.